### PR TITLE
Add document:search

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
## What does this PR do ?

This PR adds `document:search` with its units tests and the `SearchResult` class.

Sample code :

```java
import io.kuzzle.sdk.*;
import io.kuzzle.sdk.CoreClasses.Responses.Response;
import io.kuzzle.sdk.CoreClasses.SearchResult;
import io.kuzzle.sdk.Options.DocumentOptions;
import io.kuzzle.sdk.Options.KuzzleOptions;
import io.kuzzle.sdk.Options.Protocol.WebSocketOptions;
import io.kuzzle.sdk.Options.SearchOptions;
import io.kuzzle.sdk.Protocol.WebSocket;

import java.util.ArrayList;
import java.util.concurrent.ConcurrentHashMap;

public class searchDocument {
    private static Kuzzle kuzzle;

    public static void main(String[] args) {
        WebSocketOptions opts = new WebSocketOptions();
        opts.setAutoReconnect(true).setConnectionTimeout(42000);

        try {
            WebSocket ws = new WebSocket("localhost", opts);

            kuzzle = new Kuzzle(ws, (KuzzleOptions) null);
            kuzzle.connect();
            ConcurrentHashMap<String, Object> doc = new ConcurrentHashMap<>();
            doc.put("Hello", "Clarisse");
            for(int i = 0; i < 200; i+=1) {
                kuzzle.getDocumentController().create("nyc-open-data", "yellow-taxi", doc).get();
            }

            SearchOptions ops = new SearchOptions();
            ops.setFrom(0);
            ops.setSize(50);
            
            ConcurrentHashMap<String, Object> searchQuery = new ConcurrentHashMap<>();
            ConcurrentHashMap<String, Object> query = new ConcurrentHashMap<>();
            ConcurrentHashMap<String, Object> match = new ConcurrentHashMap<>();
            match.put("category", "suv");
            query.put("match", match);
            searchQuery.put("query", query);

            SearchResult result = kuzzle.getDocumentController().search("nyc-open-data", "yellow-taxi", searchQuery, ops);

            for (; result != null; result = result.next().get()) {
                System.out.println(result.hits);
            }
        }  catch (Exception e) {
            e.printStackTrace();
        }
        kuzzle.disconnect();
    }
};
```

### How should this be manually tested?

  - Step 1 :
  - Step 2 :
  - Step 3 :  
  ...

### Other changes
Correct how `refresh` option is passed and adapt tests

### CODE REVIEW

File: doc/3/controllers/document/search/index.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+type: page
+title: search
+description: Searches a document
+---
+
+# search
+
+Searches document.
+
+There is a limit to how many documents can be returned by a single search query.
```
Comment (Aschen) [inline doc/3/controllers/document/search/index.md:12]: IMHO this should be a warning

File: doc/3/controllers/document/search/index.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+That limit is by default set at 10000 documents, and you can't get over it even with the from and size pagination options.
+
+:::info
```
Comment (Aschen) [inline doc/3/controllers/document/search/index.md:15]:
```suggestion
::: info
```

File: doc/3/controllers/document/search/index.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+| ---------- | ------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `from`     | <pre>Integer</pre><br/>(`0`)    | Offset of the first document to fetch                                                                                                                                                                             |
+| `size`     | <pre>Integer</pre><br/>(`10`)   | Maximum number of documents to retrieve per page                                                                                                                                                                  |
+| `scroll`   | <pre>String</pre><br/>(`""`)    | When set, gets a forward-only cursor having its ttl set to the given value (ie `30s`; cf [elasticsearch time limits](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/common-options.html#time-units)) |
```
Comment (Aschen) [inline doc/3/controllers/document/search/index.md:68]: It's better to keep the scroll duration low
```suggestion
| `scroll`   | <pre>String</pre><br/>(`""`)    | When set, gets a forward-only cursor having its ttl set to the given value (ie `1s`; cf [elasticsearch time limits](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/common-options.html#time-units)) |
```

File: doc/3/core-classes/search-options/index.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+```java
+import io.kuzzle.sdk.Options.SearchOptions;
+```
+
+## Properties
+
+| Property | Type                  | Description                           |
+| -------- | --------------------- | ------------------------------------- |
+| `from`   | <pre>Integer</pre>    | Offset of the first document to fetch |
+| `scroll` | <pre>String</pre>     | When set, gets a forward-only cursor having its ttl set to the given value (ie `30s`; cf [elasticsearch time limits](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/common-options.html#time-units)) |
```
Comment (Aschen) [inline doc/3/core-classes/search-options/index.md:30]:
```suggestion
| `scroll` | <pre>String</pre>     | When set, gets a forward-only cursor having its ttl set to the given value (ie `1s`; cf [elasticsearch time limits](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/common-options.html#time-units)) |
```

File: doc/3/core-classes/search-result/introduction/index.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+code: true
+type: page
+title: constructor
+description: SearchResult:constructor
+order: 1
+---
+
+# Constructor
+
+This object can only be instantiated internally by this SDK, and is an easy-to-use representation of a paginated result from a [search](/sdk/java/2/core-classes/collection/search) or a [scroll](/sdk/java/2/core-classes/collection/scroll) request.
```
Comment (Aschen) [inline doc/3/core-classes/search-result/introduction/index.md:11]:
```suggestion
This object can only be instantiated internally by this SDK, and is an easy-to-use representation of a paginated result from a [search](/sdk/java/3/core-classes/collection/search) or a [scroll](/sdk/java/3/core-classes/collection/scroll) request.
```

File: doc/3/core-classes/search-result/introduction/index.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+## Properties
+
+| Property name  | Type       | Description                                               | get/set |
+| -------------- | ---------- | --------------------------------------------------------- | ------- |
+| `aggregations` | ConcurrentHashMap<String, Object>     | The result of an aggregation produced by a search request | get     |
+| `collection`   | Collection | The collection associated to this document                | get     |
+| `documents`    | ArrayList<ConcurrentHashMap<String, Object>> | An array of instantiated Document objects                 | get     |
+| `fetched`      | Integer     | The number of fetched documents so far                    | get/set |
+| `options`      | SearchOptions     | The arguments of the search/scroll request                | get     |
+| `filters`      | object     | The filters of the search request                         | get     |
```
Comment (Aschen) [inline doc/3/core-classes/search-result/introduction/index.md:24]: isn't this a ConcurrentHashMap?

File: doc/3/core-classes/search-result/next/index.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+---
+
+# next
+
+Advances through the search results and returns the next page of items.
+
+## Arguments
+
+```java
+next();
```
Comment (Aschen) [inline doc/3/core-classes/search-result/next/index.md:16]: Incomplete signature

File: doc/3/core-classes/search-result/next/snippets/scroll.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      document.put("body", body);
+      documents.add(document);
+    }
+
+    kuzzle
+      .getDocumentController()
+      .mCreate("nyc-open-data", "yellow-taxi", documents, true).get();
+
+    SearchOptions options = new SearchOptions();
+    options.setScroll("10s");
```
Comment (Aschen) [inline doc/3/core-classes/search-result/next/snippets/scroll.java:18]:
```suggestion
    options.setScroll("1s");
```

File: src/main/java/io/kuzzle/sdk/CoreClasses/SearchResult.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        searchAfter.add(this.request.get("collection").toString() + "#" + lastItem.get("_id").toString());
+      } else {
+        ConcurrentHashMap<String, Object> _source = (ConcurrentHashMap<String, Object>) lastItem.get("_source");
+        searchAfter.add(_source.get(key));
+      }
+    }
+    return nextRequest;
+  }
+
+  public SearchResult next() throws NotConnectedException, InternalException, ExecutionException, InterruptedException {
```
Comment (Aschen) [inline src/main/java/io/kuzzle/sdk/CoreClasses/SearchResult.java:105]: Calls to `next()` are not async?

Comment (Yoann-Abbes) [reply src/main/java/io/kuzzle/sdk/CoreClasses/SearchResult.java]: I've changed it, its now async

File: src/main/java/io/kuzzle/sdk/CoreClasses/SearchResult.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+      } else {
+        ConcurrentHashMap<String, Object> _source = (ConcurrentHashMap<String, Object>) lastItem.get("_source");
+        searchAfter.add(_source.get(key));
+      }
+    }
+    return nextRequest;
+  }
+
+  public CompletableFuture<SearchResult> next() throws NotConnectedException, InternalException, ExecutionException, InterruptedException {
+    System.out.println("fetched: " + this.fetched + " total: " + this.total);
```
Comment (Aschen) [inline src/main/java/io/kuzzle/sdk/CoreClasses/SearchResult.java:106]: Debug leftover

**Pull request merged**